### PR TITLE
desktop: Make clipboard support optional

### PR DESCRIPTION
**Pull request by ColinKinloch**
On Wayland ruffle uses https://github.com/YaLTeR/wl-clipboard-rs which only supports [`wlr-data-control-unstable-v1`](https://wayland.app/protocols/wlr-data-control-unstable-v1) which is not intended to be supported by all compositors.

There are options with better support https://wayland.app/protocols/primary-selection-unstable-v1 but until it's implemented in `wl-clipboard-rs` ruffle will crash for want of clipboard access.

For most platforms this MR wont effect anything. On platforms without clipboard support it'll quietly fail to set and get the clipboard.

Fixes https://github.com/ruffle-rs/ruffle/issues/15213

Mutter doesn't intend to support it: https://gitlab.gnome.org/GNOME/mutter/-/issues/524

### CODE REVIEW

File: desktop/src/backends/ui.rs [hunk 1]
```diff
@@ -136,7 +136,9 @@ impl DesktopUiBackend {
         Ok(Self {
             window,
             cursor_visible: true,
-            clipboard: Clipboard::new().context("Couldn't get platform clipboard")?,
+            clipboard: Clipboard::new()
+                .context("Couldn't get platform clipboard")
```
Comment (torokati44) [inline desktop/src/backends/ui.rs:140]: Is this going to be logged anywhere, if followed by an `.ok()` immediately? :thinking:

Comment (ColinKinloch) [reply desktop/src/backends/ui.rs:140]: I wrote this code late and I've not done much rust recently so I'm a little out of practice. Sorry.

I guess the "anyhow"s `context` call doesn't do anything in this context.

"arboard" spits out a pretty detailed error if that's sufficient, or I could check if `Clipboard::new().ok().is_none()` and warn.
```
2024-03-12T17:12:26.436744Z  WARN arboard::platform::linux: Tried to initialize the wayland data control protocol clipboard, but failed. Falling back to the X11 clipboard protocol. The error was: Unknown error while interacting with the clipboard: A required Wayland protocol (zwlr_data_control_manager_v1 version 1) is not supported by the compositor
```

Comment (evilpie) [reply desktop/src/backends/ui.rs]: You could do `Clipboard::new().inspect_err(|err| tracing::error("Failed to initialize clipboard: {err}")).ok()`

Comment (ColinKinloch) [reply desktop/src/backends/ui.rs]: Neat, thanks for the tip :)

**Review by evilpie — APPROVED** [review]

**Review by torokati44 — APPROVED** [review]
Thank you!
We really should get to the bottom of this Wayland protocol mess someday...

**Comment by evilpie** [comment]
> Thank you! We really should get to the bottom of this Wayland protocol mess someday...

Second! I really hope `wl-clipboard-rs` or `arboard` can figure this out for us :smiley:

**Comment by torokati44** [comment]
Well, of course... 😅
Or `smithay-clipboard`, that we're currently directly using.

**Comment by torokati44** [comment]
Oh, and welcome, new contributor! :wave: :smile:

**Comment by torokati44** [comment]
Speaking of - could you please raise this to https://github.com/Smithay/smithay-clipboard/issues as well, OP?
Seems like you have a better understanding of all these protocols and their implementation statuses than us...

**Comment by ColinKinloch** [comment]
> could you please raise this to smithay-clipboard

Smithays fine, I'm currently looking into whether I can use it in place of arboard.

On that, do you have any example swf files that access the clipboard?

**Comment by torokati44** [comment]
I meant the `wlr-data-control-unstable-v1` vs. `primary-selection-unstable-v1` thing.
Sounded to me like these are alternatives of each other...?

**Comment by torokati44** [comment]
> On that, do you have any example swf files that access the clipboard?

Yeah, see: https://github.com/ruffle-rs/ruffle/issues/6476#issuecomment-1071468712
Proper link in this comment, steps to reach clipboard usage in the original description.

**Comment by ColinKinloch** [comment]
In wayland the clipboard is defined in the core protocol by [`wl_data_device::selection`](https://wayland.app/protocols/wayland#wl_data_device:event:selection) and [`wl_data_device::set_selection`](https://wayland.app/protocols/wayland#wl_data_device:request:set_selection).

[`wlr-data-control-unstable-v1`](https://wayland.app/protocols/wlr-data-control-unstable-v1) seems to be for creating a "clipboard manager" which seems to be a client that is allowed to manage the clipboard.

"primary selection" refers to the ability to use middle click to paste the last highlighted text.

`egui_winit::clipboard::Clipboard` seems like an easy option as it's already in the dependency tree and wraps both arboard and smithay-clipboard. However in testing it in a flatpak sandbox it still seems to require x11 (XWayland) to access the system clipboard.

**Pull request merged**